pylops/optimization/leastsquares.py

```diff
@@ -108,6 +108,18 @@ def NormalEquationsInversion(
     """
     ncp = get_array_module(data)
 
+    # Ensure compatibility with SciPy functions
```

Comment:
Mmh I am not sure I understand this. If I understand correctly you simply check if the operator is not flattened you apply flattened, so that the solver will simply go from a multiple RHS problem to a flattened problem?

I think this is reasonable but comes with some caveats. Imagine you want to do a N-traces seismic inversion and write it as a single ``PoststackLinearModelling`` operator. You can then include a ``Laplacian`` regularizers to couple those traces. Now let's take a (perhaps stupid) case where you have 2 datasets and want to put them side by side into a N*2 matrix ``y`` and use the new option M2 to invert for a N*2 matrix  ``x``. This would flatten the operators and couple the laplacian on the first and second column of x?

I am just not sure, I am asking as you may have thought about ;)

Reply:
> Mmh I am not sure I understand this. If I understand correctly you simply check if the operator is not flattened you apply flattened, so that the solver will simply go from a multiple RHS problem to a flattened problem?

Not exactly. [The logic in the flattened part of the LinearOperator](https://github.com/PyLops/pylops/blob/c4abb6a54d5efdcf279a749b6d4acdbf184913fc/pylops/LinearOperator.py#L508-L513) does not flatten N-d array inputs. If the arrays are not shaped (N,) or (N, P), the multiplication will raise an error. This was the same behavior as the v1.x, so it is basically an easy way to ensure that your code doesn't break going from v1.x to v2. I'm tempted to make the default rely on a module-level variable but I haven't decided if that is a good idea or not. Something like `pylops.set_ndarray_multiplication(True/False)`.

In this particular function (leastsquares), I ensure all input operators are flattened = True, and then set them back. It's a bit hacky. The module-level implementation would make the code considerably easier, but globals are usually not a great idea.

Reply:
Just an addition to this. I introduced the flattened logic because of the annoying case when dims is flat i.e., dims = (N,) but dimsd is not e.g. dimsd = (M1, ..., MK). This case appears in interpolation. I previously (erroneously) assumed that flat inputs -> flat outputs. But that is clearly not the case, which means we can't rely on the shape of dims to decide whether to flatten the output or not. 

I figured the best solution was to separate the v2 behavior from v1. In v1 you always get an error when you try n-d array multiplication (unless its a properly shaped matrix). In v2, if we have a dims-shaped input, we will output a dimsd-shaped array. In the case above (dims is flat but dimsd is not), it *will* reshape the output. Taking the interpolation example, this makes sense.

Reply:
Mmh I’m confused. If you refer to Interp/Bilinear, I would say the input has dims.. if that is 1D (dims=(N,), how can you get an output that is ND?

Reply:
Regarding the first comment, so you put all operators to flattened=True so that their outputs are flat (which is what solvers need at the moment)?

Reply:
> Mmh I’m confused. If you refer to Interp/Bilinear, I would say the input has dims.. if that is 1D (dims=(N,), how can you get an output that is ND?

The issue I was having was with ``Bop.H`` in plot_bilinear.py. It takes a flat array and outputs a 2D array. Maybe I'm doing something wrong?

Reply:
> Regarding the first comment, so you put all operators to flattened=True so that their outputs are flat (which is what solvers need at the moment)?

Exactly. They will also error if the input is not flat (v1.x behavior).

Reply:
> > Mmh I’m confused. If you refer to Interp/Bilinear, I would say the input has dims.. if that is 1D (dims=(N,), how can you get an output that is ND?
> 
> The issue I was having was with `Bop.H` in plot_bilinear.py. It takes a flat array and outputs a 2D array. Maybe I'm doing something wrong?

That's correct. I was referring to the forward and that one goes from grid to random points, so the adjoint goes from random points to a grid :)

Reply:
I just committed the changes for setting the ndarray support on a module level. This also creates a nice pattern to alter behavior within functions for backward compatibility like this:

```python
from pylops import disabled_ndarray_multiplication

def old_style_func(Op, y, ...):
    x = x.ravel() # Optional, otherwise will error on non-1d arrays
    with disabled_ndarray_multiplication():
        # do stuff in v1.x way
```

Let me know if that makes sense. If you do, just resolve the conversation :)